### ISSUE TITLE
Handle futex waits in experiment sessions

### ISSUE THREAD

**Issue by rocallahan** [rr-debugger/rr#1230]
roc:
One thing I've been thinking about but forgot to mention earlier is that it's easy to imagine cases where the user-called function needs to acquire a lock which is currently held by another task. E.g., anything that calls malloc can probably hit this. In that case, we'll see the main task do a futex_wait. I think we can handle that by allowing a context switch to any other runnable task and letting that task run until it hits a syscall, and then blocking it and trying to resume execution of the main task. Of course, the other task could hit a futex_wait itself. So in general we need to keep waking up tasks and letting them run, perhaps in round-robin fashion, and allowing all futex_waits to retry, until the main task completes.

I'm not 100% sure this is going to be a real problem, so let's not solve it just yet.

cjones:

That's a good point. That could get pretty complex. For the common futex cases, we could have rr track the dag of "blocked" tasks and then emulate futex semantics (for experiment mode). Then rr could maybe know when its "target" task becomes unblocked and it can continue on normally.

A problem there is that the list of already blocked tasks, from before the switch into experiment mode / blocking futex call, won't be available to rr or the kernel. So neither just executing the syscalls, nor emulating futex in rr, is going to work in general without more effort.
